openraft/src/core/raft_core.rs

```diff
@@ -380,14 +380,24 @@ impl<C: RaftTypeConfig, N: RaftNetworkFactory<C>, S: RaftStorage<C>> RaftCore<C,
 
             // If we receive a response with a greater term, then revert to follower and abort this request.
             if let AppendEntriesResponse::HigherVote(vote) = data {
-                // TODO: there is no guarantee the response vote is greater than local. Because local vote may already
-                //       changed.
-                assert!(vote > self.engine.state.vote);
-                self.engine.state.vote = vote;
-                // TODO(xp): deal with storage error
-                self.save_vote().await.unwrap();
-                // TODO(xp): if receives error about a higher term, it should stop at once?
-                self.set_target_state(ServerState::Follower);
+                if self.does_vote_match(vote, "HigherVote during CheckIsLeader") {
+                    if let Err(e) = self.engine.handle_vote_change(&vote) {
+                        // shouldn't happend due to the check above
+                        tracing::warn!(target = display(target), "vote {vote} rejected: {e}");
+                    }
+                    if let Err(e) = self.run_engine_commands::<Entry<C>>(&[]).await.extract_fatal() {
+                        let _ = tx.send(Err(e.into()));
+                        return;
+                    }
+                } else {
+                    // simply ignore stale responses
+                    continue;
+                }
```

Comment:
`does_vote_match()` is meant to be identify events(response) from a same leader/candidate: assume a request is send by a leader with `Engine.state.vote=A`, the response will be just ignored if they are received by another leader with `Engine.state.vote=B`. E.g. leader-A sent append-entries requests, but the responses are received by leader-B(leader-A and leader-B are the same node, but in different `term`).

In this case, all the received responses are from the same `vote`.
Thus it does not have to ignore any response: `does_vote_match()` can be removed and just pass the response to `Engine.handle_vote_change()`